### PR TITLE
Add support for variadic arguments

### PR DESCRIPTION
**Pull request by tiago4orion**
The '...' syntax was added to declare functions that can receive any number of arguments. 

- A ... parameter is received as a list inside the function.

```sh
fn println(args...) {
        for a in $args {
                print($a)
        }
        print("\n")
}
```
- A ... parameter is optional
```sh
println() # Output: \n

fn a(b, c...) {
    print($b, $c...)
}
a("test") # Output: test
a() # Fail, requires at least one argument.
```
- As a special case, the ... expansion is allowed in command arguments to pass a list as multiples arguments.

```sh
# The commands below are equivalent
curl $args
curl $args...
```

This is only a first proposal, feel free to criticize.

I believe now we can change the `print` builtin to just print (no format) and then add the following functions as nash stdlib:

```sh
fn printf(fmt, args...) {
    print(format(fmt, $args...))
}

fn println(args...) {
        for arg in $args {
                print($arg)
        }
        print("\n")
}

# and so on
```

### CODE REVIEW

**Comment by katcipis** [comment]
Awesome @tiago4orion :-)

Even when I was developing print as a builtin I was not able to see an advantage on having print and printf. Do you see one ?

Just calling print("msg") already works, why have another function ?

The only advantage seems to be simpler builtin print, but it will be marginally simpler.

Building println on top of print is indeed a great idea, just need a good stdlib to add it on :+1:

**Comment by tiago4orion** [comment]
I think some way to print without formatting is useful:
For example, how we can print a raw string (eg.: string having 'format chars') ?
```sh
a = "{\"error\": \"%s\"}"
print($a) # Output: {"error": "%!s(MISSING)"}
```
For debugging purposes, it's important that we have some way to print stuff as is.
We can introduce another built-in to solve that, maybe call it `puts` (that just prints, no format and no newline), but then I would argue that `print` should be implemented in nash using this low level print:

```
fn print(fmt, args...) {
    puts(format(fmt, args...))
}
```

What do you think?

**Comment by katcipis** [comment]
@tiago4orion I see, I think I never needed this kind of stuff, but indeed it is possible. Now it makes sense :-)

**Comment by katcipis** [comment]
I'm not sure if I'm just tripping here, but the case:

```
curl $args
curl $args...
```

Made me think on a way to make the language more clear/explicit. I always found it odd to rely on the string representation of a list of strings to enable the user to pass a list of parameters for commands. I always felt that a change on the strings representation of a type would break another very important feature of the language.

Also, I always got confused why sometimes cmd $arg not worked, because it does not work when the list has other lists inside.

Enabling:

```
cmd $args...
```
As the only valid way to unpack a list of arguments seems much more clear to me, changing the strings representation of a list would not break anything, because the expansion of the list as arguments rely on a feature that is made explicitly to do that.

In conclusion, my proposal is to deprecate **cmd $arg** when $arg is a list. Commands ONLY works with string arguments, it will work with the expansion because it will expand the list to multiple strings args, on the case that it expands to something else, it will break, and it will be expected to break.

If we dont do this, lets say I have this:

```
a = ("a" ("b"))
echo $a...
```

It seems like it will work, because it will expand, and the list with strings inside are also allowed on commands, but it seems very confusing to me the possible combinations and behaviors.

Feel free to consider this just a limitation of my small skull :-)

File: ast/node_args.go [hunk 1]
```diff
@@ -111,14 +114,14 @@ func (l *ListExpr) IsEqual(other Node) bool {
...
-	if len(l.list) != len(o.list) {
+	if len(l.List) != len(o.List) {
 		return false
 	}
 
-	for i := 0; i < len(l.list); i++ {
-		if !l.list[i].IsEqual(o.list[i]) {
-			debug("%v(%s) != %v(%s)", l.list[i], l.list[i].Type(),
-				o.list[i], o.list[i].Type())
+	for i := 0; i < len(l.List); i++ {
```
Comment (katcipis) [inline ast/node_args.go]: I would avoid all those arrays accesses by using a for i, val and also creating anther variable to contain o.List[i]. The arrays accesses are copied everywhere, it is a little harder to read and easier to make a mistake (not a big thing although).

File: ast/node_fmt.go [hunk 2]
```diff
@@ -59,11 +59,18 @@ func (c *ConcatExpr) String() string {
...
-	return v.name
+	if v.IsVariadic {
+		return v.Name + "..."
+	}
+	return v.Name
 }
 
 func (i *IndexExpr) String() string {
-	return i.variable.String() + "[" + i.index.String() + "]"
+	ret := i.Var.String() + "[" + i.Index.String() + "]"
```
Comment (katcipis) [inline ast/node_fmt.go]: If you use:

```
ret := fmt.Sprintf("%s[%s]", i.Var, i.Index)
```
Won't Go automatically call the method String from your type ? (not sure)

It would be easier to read, and less stuff to write :-)

File: internal/sh/fncall.go [hunk 3]
```diff
@@ -41,22 +46,79 @@ func NewUserFn(name string, parent *Shell) (*UserFn, error) {
...
 func (fn *UserFn) SetArgs(args []sh.Obj) error {
-	if len(fn.argNames) != len(args) {
-		return errors.NewError("Wrong number of arguments for function %s. Expected %d but found %d",
+	var (
+		isVariadic      bool
+		countNormalArgs int
+	)
+
+	for i := 0; i < len(fn.argNames); i++ {
+		argName := fn.argNames[i]
```
Comment (katcipis) [inline internal/sh/fncall.go]: Why not:
```
for i, argName := 0; i < len(fn.argNames); i++ {
```

Not C enough ? =)

File: internal/sh/fncall.go [hunk 4]
```diff
@@ -41,22 +46,79 @@ func NewUserFn(name string, parent *Shell) (*UserFn, error) {
...
+			// there's only a variadic (optional) argument
+			// and user supplied no argument...
+			// then only initialize the variadic variable to
+			// empty list
+			fn.Setvar(fn.argNames[0].Name, sh.NewListObj([]sh.Obj{}))
+			return nil
+		}
+	}
+
+	var i int
```
Comment (katcipis) [inline internal/sh/fncall.go]: To a variable that is even used outside the loop scope **i** seems to be a too short name, broader scope demands improved naming

**Comment by tiago4orion** [comment]
I'm not sure I understand what you mean here:
> I always found it odd to rely on the string representation of a list of strings to enable the user to pass a list of parameters for commands. I always felt that a change on the strings representation of a type would break another very important feature of the language.

When you pass a list to a command, as in `cmd $arg`, it does not rely in `arg.String()`, but it proper iterate the list and check if it's composed of strings. The code is here:
https://github.com/NeowayLabs/nash/blob/master/internal/sh/cmd.go#L79

> Also, I always got confused why sometimes cmd $arg not worked, because it does not work when the list has other lists inside.

Well, in every operating system (that I know) arguments are just strings (or null-terminated byte array in unix-like OS). That's why nash aborts if the list isn't a list of strings.. not a nash limitation of the string representation of a list. But maybe I misunderstand your assumption.

File: internal/sh/shell.go [hunk 5]
```diff
@@ -1476,17 +1481,57 @@ func (shell *Shell) evalIndexedVar(indexVar *ast.IndexExpr) (sh.Obj, error) {
...
+		return nil, errors.NewEvalError(shell.filename, indexVar.Var,
+			"Invalid indexing of non-list variable: %s (%+v)", v.Type(), v)
+	}
+
+	indexNum, err := shell.evalIndex(indexVar.Index)
+	if err != nil {
+		return nil, err
+	}
+
+	vlist := v.(*sh.ListObj)
```
Comment (katcipis) [inline internal/sh/shell.go]: Allowing panic here is desired ? (I imagine that in this case it is, just to be sure)

File: internal/sh/shell.go [hunk 6]
```diff
@@ -1499,22 +1544,55 @@ func (shell *Shell) evalVariable(a ast.Expr) (sh.Obj, error) {
 	}
 
 	vexpr := a.(*ast.VarExpr)
-	varName := vexpr.Name()
+	varName := vexpr.Name
 
-	if v, ok = shell.Getvar(varName[1:]); !ok {
+	if value, ok = shell.Getvar(varName[1:]); !ok {
```
Comment (katcipis) [inline internal/sh/shell.go]: are there guarantees that the varName will always have a size > 1 ?

File: internal/sh/shell.go [hunk 7]
```diff
@@ -1499,22 +1544,55 @@ func (shell *Shell) evalVariable(a ast.Expr) (sh.Obj, error) {
 	}
 
 	vexpr := a.(*ast.VarExpr)
```
Comment (katcipis) [inline internal/sh/shell.go]: Another panic. I understand that these are very exceptional situations, but I'm afraid that if they happen the panic message will not be very clear on what went wrong (although the stack usually is enough to track down, if the person knows the code well). I honestly do not like the idea of the language tossing up panics from Go. It gives more work but it seems nicer to provide nice nash errors, even for fatal stuff (seems like we are leaking Go if something goes wrong on our code).

File: internal/sh/shell.go [hunk 8]
```diff
@@ -1525,6 +1603,79 @@ func (shell *Shell) evalExprs(exprs []ast.Expr) ([]sh.Obj, error) {
...
+		}
+
+		ret = append(ret, objs...)
+	}
+
+	return ret, nil
+}
+
+func (shell *Shell) evalArgExpr(expr ast.Expr) ([]sh.Obj, error) {
+	switch expr.Type() {
```
Comment (katcipis) [inline internal/sh/shell.go]: Instead of having this switch case would it not be better if each ast.Expr had a Eval method (interface level). Besides removing the switch you would also remove the casts (which are also not very desirable).

Doing like that, every type concentrates the code to handle that type, the way it is done today handling of each type is spread through the code on different switch cases (perhaps on this case is just this one, but they tend to multiply).

File: internal/sh/shell.go [hunk 9]
```diff
@@ -1525,6 +1603,79 @@ func (shell *Shell) evalExprs(exprs []ast.Expr) ([]sh.Obj, error) {
...
+			return []sh.Obj{objs[0]}, nil
+		}
+	}
+
+	return nil, errors.NewEvalError(shell.filename,
+		expr, "Failed to eval expression: %+v", expr)
+}
+
 func (shell *Shell) evalExpr(expr ast.Expr) (sh.Obj, error) {
 	switch expr.Type() {
```
Comment (katcipis) [inline internal/sh/shell.go]: Yeah, and the switch case multiples =)

**Comment by katcipis** [comment]
Sorry, once we talked about that and I understood that the string representation of a list of strings matches with the arguments required for the command (printing a list of strings generates a space separated sequence of strings, matches with what the command expects), and that is why it worked. Even tough, I'm not sure if it good to have these two ways, it made sense when we did not have vargs and expansion, now that we have it seems a little confusing to me, but it is not a big deal.

File: internal/sh/shell_test.go [hunk 10]
```diff
@@ -1216,7 +1216,7 @@ func TestExecutePipe(t *testing.T) {
 	}
 
 	expectedOutput := "hello"
-	actualOutput   := string(stdout.Bytes())
+	actualOutput := string(stdout.Bytes())
```
Comment (katcipis) [inline internal/sh/shell_test.go]: bytes.Buffer already has a String method.

Comment (katcipis) [inline internal/sh/shell_test.go]: For this kinda of e2e testing it seems practical to create a small package that execs a command and already collects output in memory, like Go's Cmd Output() and CombinedOuptut()

**Comment by tiago4orion** [comment]
Yeah, I'm inclined to deprecate the old way also, but today this feature is used everywhere... a big break change.
We can deprecate it now, and remove when we launch the 1.0

File: internal/sh/shell_test.go [hunk 11]
```diff
@@ -2301,3 +2301,73 @@ world`)
...
+		},
+		{
+			desc: "the first argument isn't optional",
+			execStr: `fn a(b, c...) {
+    print($b, $c...)
+}
+a()`,
+			expectedErr: "Wrong number of arguments for function a. Expected at least 1 arguments but found 0",
+		},
+	} {
```
Comment (katcipis) [inline internal/sh/shell_test.go]: It seems to be missing a test for the case of a empty list being expanded (I'm curious on the behavior =)).

File: parser/parse.go [hunk 12]
```diff
@@ -777,7 +792,13 @@ func (p *Parser) parseList(tok *scanner.Token) (ast.Node, error) {
 	}
 
 	p.ignore()
-	return ast.NewListExpr(lit.FileInfo, values), nil
+
+	var isVariadic bool
+	if p.peek().Type() == token.Dotdotdot {
```
Comment (katcipis) [inline parser/parse.go]: auheuhaeuhuaheuaheuaheuahe

dotdotdot =)

File: parser/parse.go [hunk 13]
```diff
@@ -1017,24 +1038,27 @@ func (p *Parser) parseFnArgs() ([]string, error) {
 
 	for {
```
Comment (katcipis) [inline parser/parse.go]: Living dangerously lol

File: parser/parse.go [hunk 14]
```diff
@@ -1388,7 +1406,7 @@ func (p *Parser) parseFor(it scanner.Token) (ast.Node, error) {
 	if (it.Type() == token.Ident || it.Type() == token.Variable) && next.Type() == token.LParen {
```
Comment (katcipis) [inline parser/parse.go]: Smells like a switch case disguised as ifs =)

File: ast/node_args.go [hunk 15]
```diff
@@ -85,19 +85,22 @@ func (i *IntExpr) IsEqual(other Node) bool {
...
 		NodeType: NodeListExpr,
 		FileInfo: info,
 
-		list: values,
+		List:       values,
+		IsVariadic: variadic,
 	}
 }
 
-func (l *ListExpr) List() []Expr { return l.list }
```
Comment (tiago4orion) [inline ast/node_args.go]: I'm following your suggestion and removing the getters whenever possible =)

File: ast/node_args.go (hunk 1 above)
Comment (tiago4orion) [inline ast/node_args.go]: :+1:

File: spec.ebnf [hunk 16]
```diff
@@ -59,7 +59,7 @@ returnDecl = "return" [ ( variable | stringLit | list | fnInv ) ] .
 fnInv = ( variable | identifier ) "(" fnArgValues ")" .
 
 fnArgValues = { fnArgValue [ "," ] } .
-fnArgValue  = [ stringLit | stringConcat | list | variable | fnInv ] .
+fnArgValue  = [ stringLit | stringConcat | list | (variable [ "..." ]) | (list [ "..." ]) fnInv ] .
```
Comment (katcipis) [inline spec.ebnf]: I think there is a way to build a BNF that expresses that the variadic arg is only allowed as the last param (not completely sure, it is late, but if you agree I can give it a shot =))

File: ast/node_fmt.go (hunk 2 above)
Comment (tiago4orion) [inline ast/node_fmt.go]: yes, it will call our String() method. I can change, but sometimes I'm worried about the complexity of fmt package =/ But I've brain damage kkk

Comment (katcipis) [inline ast/node_fmt.go]: @tiago4orion take it easy on all this anxiety, it will give you headaches =P

File: internal/sh/builtin/append.go [hunk 17]
```diff
@@ -9,40 +9,41 @@ import (
...
+func (appendfn *appendFn) ArgNames() []sh.FnArg {
+	return []sh.FnArg{
+		sh.NewFnArg("list", false),
+		sh.NewFnArg("value...", true), // variadic
+	}
 }
 
 func (appendfn *appendFn) Run(in io.Reader, out io.Writer, err io.Writer) ([]sh.Obj, error) {
-	newobj := append(appendfn.obj, appendfn.arg)
+	newobj := append(appendfn.obj, appendfn.args...)
```
Comment (tiago4orion) [inline internal/sh/builtin/append.go]: forgot to document and test this change :disappointed: 
Now `append` builtin supports multiple arguments.

Comment (katcipis) [inline internal/sh/builtin/append.go]: I noticed that you changed that on the beginning of the review, but got lost on other stuff and forgot to mention that I did not found the test anywhere =P

File: internal/sh/fncall.go (hunk 3 above)
Comment (tiago4orion) [inline internal/sh/fncall.go]: o.O

Comment (katcipis) [inline internal/sh/fncall.go]: Sorry, I gave a bugged example of what I had in mind =P

File: internal/sh/shell.go (hunk 5 above)
Comment (tiago4orion) [inline internal/sh/shell.go]: Few lines above we check the type of `v`, then here it should be a ListObj. If it's not the case, then the shell is at an inconsistent state, right? This seems to be a good candidate to a panic because it's not safe to continue (a bug was triggered or a race happened somewhere). What do you think? And if is better to validate here, what should I do?

File: internal/sh/shell.go (hunk 6 above)
Comment (tiago4orion) [inline internal/sh/shell.go]: A variable always start with a $ sign

File: internal/sh/shell.go (hunk 5 above)
Comment (katcipis) [inline internal/sh/shell.go]: I understand that it is not safe to continue, but usually errors that you produce directly on the code will give a better hint of what can be wrong than a generic Go panic of invalid casting. But if in this case you can't come up with anything better it is ok =)

File: internal/sh/shell.go (hunk 6 above)
Comment (katcipis) [inline internal/sh/shell.go]: This array has been validated for this before on another part of the code ? That part of the code is completely bug/bullet proof ? If it is not, which kind of error will explode ? Will it be easy to understand/debug ?

File: internal/sh/fncall.go (hunk 3 above)
Comment (tiago4orion) [inline internal/sh/fncall.go]:
```
for i, argName := range fn.argNames {}
```
?
:joy: 
what's the problem with `initializations; cond; increment` ? too much typing?
I can change, but I use this a lot..

File: internal/sh/fncall.go (hunk 4 above)
Comment (tiago4orion) [inline internal/sh/fncall.go]: It's not used in a broader scope.. It's used in the function scope and the meaning is the same: indexing

From the `PikeStyle`:
_Indices are just notation, so treat them as such._

In this case I cannot think of a better name..

File: internal/sh/shell.go (hunk 7 above)
Comment (tiago4orion) [inline internal/sh/shell.go]: I don't think so.. The interpreter is just eval of data structures created by parser and every node of AST has a proper type information for we rely upon.. otherwise every eval will be a mess of _if else logic_ to catch bugs in previous stages of the interpreter. In this stage (the evaluator) the only thing we can say is that: _parser or scanner has a bug_. And I think this is useless as a stacktrace to the user. For we (maintainers) the stack is much more useful.

The issue is that our `ast` package is exported, then users can generate ASTs and feed the interpreter directly (code generation, and so on) and then this way the evaluator will trigger panics if something do not conform.
To solve that we can work in the `ast` package to add more guarantees, validating nodes created by users or parser, then this panic will never be triggered here, maybe this will require changing the direct access to the struct members to setters and getters again...

Validating nodes in the evaluator will confuse the logic and compromise readability (I think), but I'm open to suggestions.

File: internal/sh/shell.go (hunk 6 above)
Comment (tiago4orion) [inline internal/sh/shell.go]: The scanner only emits `token.Variable` when it found a '$' followed by an identifier:
https://github.com/NeowayLabs/nash/blob/master/scanner/lex.go#L276

The again this code will fail only if the `scanner` is buggy generating `token.Variable` for wrong tokens.

Comment (tiago4orion) [inline internal/sh/shell.go]: We can take it simple or make it terrible verbose and spaguetti...
Better to guarantee mistakes by adding tests than by adding never reached lines of code.

File: internal/sh/shell.go (hunk 8 above)
Comment (tiago4orion) [inline internal/sh/shell.go]: you're arguing to mix the interpreter with the AST? I don't know if this is a good design decision...
To eval a node we need the environment (the set of variables, functions, bindings and scopes created) of the shell, then the node will need the shell, but the shell need to import the nodes to eval... There's cyclic dependency here apart from being bad giving eval powers to each node type.

Comment (tiago4orion) [inline internal/sh/shell.go]: > Doing like that, every type concentrates the code to handle that type, the way it is done today handling of each type is spread through the code on different switch cases (perhaps on this case is just this one, but they tend to multiply).

Before this PR, nash had only one `evalExpr` and then this switch case occurred in only one place (one function).
But now with variadic arguments I need to evaluate a expression a bit different if it's in a function argument.. Because '...' cannot be used in other places.
My first try was adding a new boolean flag to `evalExprs` and making it handle both cases.. but the code looked messy and then replicated a bit of the logic to make the code more explicit of what it is being doing.

File: internal/sh/shell.go [hunk 18]
```diff
@@ -1884,42 +2033,13 @@ func (shell *Shell) executeIfNotEqual(n *ast.IfNode) ([]sh.Obj, error) {
 	return nil, nil
 }
 
-func (shell *Shell) executeFn(fn sh.Fn, nodeArgs []ast.Expr) ([]sh.Obj, error) {
```
Comment (tiago4orion) [inline internal/sh/shell.go]: found an unused function... o.O

File: internal/sh/shell_test.go (hunk 10 above)
Comment (tiago4orion) [inline internal/sh/shell_test.go]: > bytes.Buffer already has a String method.

Thanks

Comment (vitorarins) [inline internal/sh/shell_test.go]: Yeah, I was thinking about this. In my last PR I didn't want to add another `execTestCase...` function just for this. But I think the idea of creating a package is great. :beers:

File: internal/sh/builtin/append.go (hunk 17 above)
Comment (tiago4orion) [inline internal/sh/builtin/append.go]: :+1:

File: internal/sh/shell_test.go (hunk 10 above)
Comment (tiago4orion) [inline internal/sh/shell_test.go]: The pkg will need to collect stdout and stderr separately. We're missing to test the stderr of several tests.. looking only in the stdout is not good.. if something not expected are going to stderr we're not tracking.

Comment (tiago4orion) [inline internal/sh/shell_test.go]: I think @katcipis started such `cmd` in their last PR, no?

File: internal/sh/shell_test.go (hunk 11 above)
Comment (tiago4orion) [inline internal/sh/shell_test.go]: true
:+1:

File: parser/parse.go (hunk 14 above)
Comment (tiago4orion) [inline parser/parse.go]: hey, I did not touch this line =) kkkk
But the boolean expression is: (A || B) && C 
I'm not seeing a switch here, but I must be blindly wrong.

File: spec.ebnf (hunk 16 above)
Comment (tiago4orion) [inline spec.ebnf]: I'm fine. Because of this assumption I validated in the interpreter also with the vararg is the last argument. But making it explicit in the BNF, then we can validate this only in the parser.

File: parser/parse.go (hunk 12 above)
Comment (tiago4orion) [inline parser/parse.go]: It escalated from "Getting dot-dot right" =P
http://plan9.bell-labs.com/sys/doc/lexnames.html

File: internal/sh/fncall.go (hunk 3 above)
Comment (katcipis) [inline internal/sh/fncall.go]: No problem at all, it is only odd when you do this:

```
for i := 0; i < len(fn.argNames); i++ {
 		argName := fn.argNames[i]
```

The first thing you do is to create a var to hold the value of fn.argNames[i]. It is smaller and more efficient to just write:

```
for i, argName :=  range fn.argNames{
}
```

I think this was the purpose of the range idiom for arrays/slices =). But to be honest I was more messing around, there is no problem at all with using the increment stuff =P

File: internal/sh/fncall.go (hunk 4 above)
Comment (katcipis) [inline internal/sh/fncall.go]: Broader scope = outside the scope of the for loop, not necessarily a package scope =P. But if in this case it is just an index, that is ok, it just sounded a little odd

File: internal/sh/shell.go (hunk 7 above)
Comment (katcipis) [inline internal/sh/shell.go]: Well, thinking that error handling compromises readability is the origin of exceptions =P. If the programming error is really severely stupid perhaps a panic is better, I don't know nash design well enough, sorry

File: internal/sh/shell.go (hunk 6 above)
Comment (katcipis) [inline internal/sh/shell.go]: Did not know it was validated before, just to confirm =)

File: internal/sh/shell.go (hunk 8 above)
Comment (katcipis) [inline internal/sh/shell.go]: Did not noticed that I would mix the interpreter and the AST, sorry =). Would have to get more acquainted with the code/design to make a better proposal. Sorry for the noise

File: internal/sh/shell.go (hunk 18 above)
Comment (katcipis) [inline internal/sh/shell.go]: I use gounused on some projects and it works very well

File: internal/sh/shell_test.go (hunk 10 above)
Comment (katcipis) [inline internal/sh/shell_test.go]: @tiago4orion yep I did =). I don't get the stderr yet (don't used it)

File: parser/parse.go (hunk 14 above)
Comment (katcipis) [inline parser/parse.go]: I was looking at the "else ifs". It seems to be writable as a switch, and I usuall prefer switches than if elses =P.

All nash code has been written by you =P

File: spec.ebnf (hunk 16 above)
Comment (katcipis) [inline spec.ebnf]: Yeah I noticed the checks for the last param, I was curious on how would be the BNF =)

File: internal/sh/shell.go (hunk 7 above)
Comment (tiago4orion) [inline internal/sh/shell.go]: there's no `error handling` here.. because there's no `error`.
If we substitute this to use exceptions (if go supported them) then still no error will be handled... the only thing we can do here is abort the shell...

But.. If you think stopping using the Type() method of the node and instead use only golanf type assertions and reflection to make the type switchs is a good use case.. then I must agree that we *do* need to treat each type assertion (because only the runtime has the truth.. an internal `type` member).

Comment (katcipis) [inline internal/sh/shell.go]: If there is no error then it is ok to not handle errors =), just brainstorming

Comment (tiago4orion) [inline internal/sh/shell.go]: no problem bro. Good to discuss that =)

**Comment by tiago4orion** [comment]
This PR introduced a bug when parsing the execution of commands in the current directory:
```sh
λ> ./rkt
ERROR: <stdin line 13>:2:0: Unexpected symbol 'EOF'
```
The fix was:
https://github.com/NeowayLabs/nash/pull/220/commits/f66ec2c71f33445f979a0810fabb99dad4d9f3c9#diff-32d9b0cbc79bb412d52c5ebf6cd1673eR367

I don't know if you guys know it, but nash lexer emits automatic semicolons for us. Then, the code below:

```sh
fn print(arg) {
    echo $arg
}

_, status <= stat -d /etc
if $status == "0" {
    print("/etc exists")
}
```
The parser sees the code below:
```sh
fn print(arg) {
    echo $arg;
}

_, status <= stat -d /etc;
if $status == "0" {
    print("/etc exists");
}
```
The problem was: When scanning the '...', if only one or two dots are found, the lexer emits a token.Arg containing ".." but I forgot to handle the semicolon insertion in this case and then the parser failed because the parser is *very* strict regarding semicolons (like C or Java).
There must be a way to avoid handling token.Arg in two places, but I was not able to get a simple way to do that for now =(

**Comment by katcipis** [comment]
Oops, missed the bug that has been found, sorry

**Comment by tiago4orion** [comment]
relax, impossible to catch all my mistakes in one review =)

**Comment by tiago4orion** [comment]
Added some missing tests and cosmetic changes.
/cc @vitorarins @katcipis

File: internal/sh/builtin/append.go [hunk 19]
```diff
@@ -37,7 +37,8 @@ func (appendfn *appendFn) SetArgs(args []sh.Obj) error {
 
 	obj := args[0]
 	if obj.Type() != sh.ListType {
-		return errors.NewError("appendfn expects a list as first argument, but a %s[%s] was provided", obj, obj.Type())
+		return errors.NewError("appendfn expects a list as first argument, but a %s was provided",
```
Comment (katcipis) [inline internal/sh/builtin/append.go]: Wouldn't be better to call "append" instead of "appendfn" ? appendfn does not exist on nash, it may confuse the user.

File: internal/sh/builtin/append_test.go [hunk 20]
```diff
@@ -7,45 +7,106 @@ import (
...
-		return
+func TestAppend(t *testing.T) {
+	for _, tc := range []testcase{
+		{
+			name:        "at least two arguments",
+			code:        `append()`,
+			expectedErr: "<interactive>:1:0: appendfn expects at least two arguments",
+		},
+		{
+			name:        "at least two arguments",
```
Comment (katcipis) [inline internal/sh/builtin/append_test.go]: repeteated name ? Not sure it is a good idea. Perhaps "no arguments fails" and "one argument fails'

File: ast/node_args.go [hunk 21]
```diff
@@ -118,10 +118,11 @@ func (l *ListExpr) IsEqual(other Node) bool {
 		return false
 	}
 
-	for i := 0; i < len(l.List); i++ {
-		if !l.List[i].IsEqual(o.List[i]) {
-			debug("%v(%s) != %v(%s)", l.List[i], l.List[i].Type(),
-				o.List[i], o.List[i].Type())
+	for i, val := range l.List {
+		oval := o.List[i]
```
Comment (katcipis) [inline ast/node_args.go]: :+1:

File: ast/node_fmt.go [hunk 22]
```diff
@@ -66,7 +67,7 @@ func (v *VarExpr) String() string {
 }
 
 func (i *IndexExpr) String() string {
-	ret := i.Var.String() + "[" + i.Index.String() + "]"
+	ret := fmt.Sprintf("%s[%s]", i.Var, i.Index)
```
Comment (katcipis) [inline ast/node_fmt.go]: :+1:

File: internal/sh/fncall.go [hunk 23]
```diff
@@ -58,8 +58,7 @@ func (fn *UserFn) SetArgs(args []sh.Obj) error {
 		countNormalArgs int
 	)
 
-	for i := 0; i < len(fn.argNames); i++ {
-		argName := fn.argNames[i]
+	for i, argName := range fn.argNames {
```
Comment (katcipis) [inline internal/sh/fncall.go]: :+1:
